src/Microsoft.AspNet.Razor/TagHelpers/TagHelperHelper.cs

```diff
@@ -0,0 +1,105 @@
+        /// <paramref name="name"/>. <c>null</c> if the attribute is not bound.
+        /// </returns>
+        /// <remarks>
+        /// This method cannot use Reflection to generally determine the most specific <see cref="Type"/> when
+        /// multiple tag helpers bind the attribute. However when it can determine an attribute is bound to a
+        /// <see cref="string"/> property or dictionary value, it always returns <c>"System.String"</c>. Otherwise it
+        /// returns the full <see cref="Type"/> name for the first binding it finds, if any.
+        /// </remarks>
+        public static string GetPropertyType(
+            [NotNull] string name,
+            [NotNull] IEnumerable<TagHelperDescriptor> descriptors)
+        {
+            // Find first TagHelperAttributeDescriptor matching this name.
+            var firstBoundAttribute = descriptors
+                .SelectMany(descriptor => descriptor.Attributes)
+                .FirstOrDefault(attribute =>
+                    string.Equals(attribute.Name, name, StringComparison.OrdinalIgnoreCase) ||
+                    (attribute.Prefix != null &&
+                     name.StartsWith(attribute.Prefix, StringComparison.OrdinalIgnoreCase)));
+            if (firstBoundAttribute == null)
+            {
+                return null;
+            }
+
+            if (!IsBoundNonStringAttribute(name, descriptors))
+            {
+                // Must have been a string attribute since we know it is bound. Use that type.
+                return typeof(string).FullName;
+            }
+
+            if (string.Equals(firstBoundAttribute.Name, name, StringComparison.OrdinalIgnoreCase))
+            {
+                return firstBoundAttribute.TypeName;
+            }
+
+            // Attribute is bound and doesn't match Name. Must be a Prefix match.
+            return firstBoundAttribute.PrefixedValueTypeName;
+        }
+
+        /// <summary>
+        /// Determines whether an attribute named <paramref name="name"/> is bound exclusively to
+        /// non-<see cref="string"/> tag helper properties.
+        /// </summary>
+        /// <param name="name">Name of the attribute.</param>
+        /// <param name="descriptors">
+        /// Collection of <see cref="TagHelperDescriptor"/>s targeting the containing element in the Razor source.
+        /// </param>
+        /// <returns>
+        /// <c>true</c> if the attribute named <paramref name="name"/> is bound and no associated property or
+        /// dictionary value (in the case of a <see cref="TagHelperAttributeDescriptor.Prefix"/> match) has
+        /// <see cref="Type"/> <see cref="string"/>. <c>false</c> otherwise e.g. if the property is not bound.
+        /// </returns>
+        public static bool IsBoundNonStringAttribute(
+            [NotNull] string name,
+            [NotNull] IEnumerable<TagHelperDescriptor> descriptors)
```

Comment:
Previous code that determined if an attribute was bound to a `string` property could get very confused when multiple tag helpers bound the same name.  Of course the older code also did not handle prefix matches.